Buildings/Controls/OBC/Utilities/PIDWithInputGains.mo

```diff
@@ -0,0 +1,749 @@
+    annotation (Placement(transformation(extent={{160,-160},{180,-140}})));
+
+  block Derivative
+    "Block that approximates the derivative of the input"
+    parameter Real y_start=0
+      "Initial value of output (= state)"
+      annotation (Dialog(group="Initialization"));
+    parameter Real Nd(
+      min=100*CDL.Constants.eps)=10
+      "The higher Nd, the more ideal the derivative block";
+    CDL.Interfaces.RealInput u "Connector of Real input signal"
+      annotation (Placement(transformation(extent={{-140,-20},{-100,20}})));
+     CDL.Interfaces.RealInput k "Connector of gain signal"
+      annotation (Placement(transformation(extent={{-140,40},{-100,80}})));
+     CDL.Interfaces.RealInput Td "Connector of time constant signal for the derivative term"
+      annotation (Placement(transformation(extent={{-140,-40},{-100,-80}})));
+    CDL.Interfaces.RealOutput y "Connector of Real output signal"
+      annotation (Placement(transformation(extent={{100,-20},{140,20}})));
+    output Real x
+      "State of block";
+
+  initial equation
+    if abs(k*Td) < 1E-17 then
+      x=u;
+    else
+      x=u-y_start/k/Nd;
+    end if;
+
+  equation
+    der(x)=
+      if abs(k*Td) < 1E-17 then
+        0
+      else
+        (u-x)*Nd/Td;
+    y=if abs(k*Td) < 1E-17 then
+        0
+      else
+        (k*Nd)*(u-x);
+    annotation (
+      defaultComponentName="der",
+      Documentation(
+        info="<html>
+<p>
+This blocks defines the transfer function between the
+input <code>u</code> and the output <code>y</code>
+as <i>approximated derivative</i>:
+</p>
+<pre>
+             k * s
+     y = ------------ * u
+            T * s + 1
+</pre>
+<p>
+If <code>k=0</code>, the block reduces to <code>y=0</code>.
+</p>
+</html>",
+        revisions="<html>
+<ul>
+<li>
+May 9, 2022, by Sen Huang:<br/>
+Made the gain and the time constant as inputs</code>.<br/>
+This is for
+<a href=\"https://github.com/lbl-srg/modelica-buildings/issues/2993\">issue 2993</a>.
+</li>
+<li>
+November 12, 2020, by Michael Wetter:<br/>
+Reformulated to remove dependency to <code>Modelica.Units.SI</code>.<br/>
+This is for
+<a href=\"https://github.com/lbl-srg/modelica-buildings/issues/2243\">issue 2243</a>.
+</li>
+<li>
+August 7, 2020, by Michael Wetter:<br/>
+Moved to protected block in PID controller because the derivative block is no longer part of CDL.
+</li>
+<li>
+April 21, 2020, by Michael Wetter:<br/>
+Removed option to not set the initialization method or to set the initial state.
+The new implementation only allows to set the initial output, from which
+the initial state is computed.
+<br/>
+This is for
+<a href=\"https://github.com/lbl-srg/modelica-buildings/issues/1887\">issue 1887</a>.
+</li>
+<li>
+March 2, 2020, by Michael Wetter:<br/>
+Changed icon to display dynamically the output value.
+</li>
+<li>
+March 24, 2017, by Jianjun Hu:<br/>
+First implementation, based on the implementation of the
+Modelica Standard Library.
+</li>
```

Comment:
No need to reproduce all these revisions, simply state what your implementation is based on.

Reply:
Thanks, I deleted the records of old revisions.